src/main/java/com/kobylynskyi/graphql/codegen/scala/ScalaGraphQLTypeMapper.java

```diff
@@ -0,0 +1,127 @@
+     */
+    @Override
+    public String wrapSuperTypeIntoList(MappingContext mappingContext, String type) {
+        return getGenericsString(mappingContext, SCALA_UTIL_LIST, "_ <: " + type);
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public String wrapApiReturnTypeIfRequired(MappingContext mappingContext,
+                                              NamedDefinition namedDefinition,
+                                              String parentTypeName) {
+        String computedTypeName = namedDefinition.getJavaName();
+        if (parentTypeName.equalsIgnoreCase(GraphQLOperation.SUBSCRIPTION.name()) &&
+                Utils.isNotBlank(mappingContext.getSubscriptionReturnType())) {
+            // in case it is subscription and subscriptionReturnType is set
+            return getGenericsString(mappingContext, mappingContext.getSubscriptionReturnType(), computedTypeName);
+        }
+
+        if (Boolean.TRUE.equals(mappingContext.getUseOptionalForNullableReturnTypes()) && !namedDefinition.isMandatory()) {
+            if (!computedTypeName.startsWith(SCALA_UTIL_LIST) && !computedTypeName.startsWith(JAVA_UTIL_LIST)) {
+                // wrap the type into java.util.Optional (except java list and scala list)
+                computedTypeName = getGenericsString(mappingContext, SCALA_UTIL_OPTIONAL, computedTypeName);
+            }
+        }
+
+        if (computedTypeName.startsWith(SCALA_UTIL_LIST) &&
+                Utils.isNotBlank(mappingContext.getApiReturnListType())) {
+            // in case it is query/mutation, return type is list and apiReturnListType is set
+            return computedTypeName.replace(SCALA_UTIL_LIST, mappingContext.getApiReturnListType());
+        }
+        if (Utils.isNotBlank(mappingContext.getApiReturnType())) {
+            // in case it is query/mutation and apiReturnType is set
+            return getGenericsString(mappingContext, mappingContext.getApiReturnType(), computedTypeName);
+        }
+        return getTypeConsideringPrimitive(mappingContext, namedDefinition, computedTypeName);
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public boolean isPrimitive(String scalaType) {
+        return isScalaPrimitive(scalaType);
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public String getGenericsString(MappingContext mappingContext, String genericType, String typeParameter) {
+        return String.format("%s[%s]", genericType, typeParameter);
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public String mapDirectiveArgumentValue(MappingContext mappingContext, Argument dirArg, String argumentValueFormatter) {
+        return valueMapper.map(mappingContext, dirArg.getValue(), null, argumentValueFormatter);
+    }
+
+    @Override
+    public boolean addModelValidationAnnotationForType(String possiblyPrimitiveType) {
+        return !JavaGraphQLTypeMapper.isJavaPrimitive(possiblyPrimitiveType); // reconsider
```

Comment:
isPrimitive ?

Reply:
Yes! That's what was present in the original non-refactored code.
Scala tests were failing because of this.

Reply:
No problem at the moment, I'll optimize it later.(Annotations on Scala primitive types should be removed from the test and code.)